### PR TITLE
fix[pytest]: Default JSON encoding for non-serializable types to __repr__

### PR DESCRIPTION
**Pull request by Yun-Kim**
## Description
In cases of parameterized pytest cases, the current pytest plugin attempts to JSON-encode non-serializable Python objects to include their attributes via the `__dict__` attribute, and defaulting to `__repr__` if the `__dict__` is empty or non-existent. However, complex Python objects like the `MagicMock()` class can sometimes contain circular references to itself (for example, including a reference to its child, itself which also includes a reference to its parent) and cause the following error:
```ValueError: Circular reference detected```

To fix this and the likely similar issues in the future, the pytest plugin now defaults straight to the string representation `__repr__` of a non-serializable Python object when attempting to JSON-encode pytest test parameters.

### CODE REVIEW

File: tests/contrib/pytest/test_pytest.py [hunk 1]
```diff
@@ -123,18 +144,13 @@ def item_param():
...
-                    pytest.param({"a": A("test_name", "value"), "b": [1, 2, 3]}, marks=pytest.mark.skip),
-                    pytest.param([1, 2, 3], marks=pytest.mark.skip),
-                    pytest.param(item_param, marks=pytest.mark.skip)
-                ]
+            'item',
+            [
+                pytest.param(A("test_name", "value"), marks=pytest.mark.skip),
+                pytest.param(A("test_name", A("inner_name", "value")), marks=pytest.mark.skip),
+                pytest.param(item_param, marks=pytest.mark.skip),
+                pytest.param({"a": A("test_name", "value"), "b": [1, 2, 3]}, marks=pytest.mark.skip),
```
Comment (Kyle-Verhoog) [inline tests/contrib/pytest/test_pytest.py:152]: Does one of these cases have a circular reference to reproduce the issue?

Comment (Yun-Kim) [reply tests/contrib/pytest/test_pytest.py:152]: Not yet, will add one though 😄

File: tests/contrib/pytest/test_pytest.py [hunk 2]
```diff
@@ -123,18 +145,14 @@ def item_param():
...
+                pytest.param(MagicMock(value=MagicMock()), marks=pytest.mark.skip),
```
Comment (Kyle-Verhoog) [inline tests/contrib/pytest/test_pytest.py:154]: If this is the case added to reproduce a circular reference - can we be more explicit. MagicMock could change its internals at any time which might remove the circular reference. Can we create our own simple class that contains a circular reference instead?

Comment (jd) [reply tests/contrib/pytest/test_pytest.py:154]: That does not seem circular to me.
I think we miss such a case?

Comment (Yun-Kim) [reply tests/contrib/pytest/test_pytest.py:157]: Yes, `MagicMock` objects that contain `MagicMock` attributes result in a circular reference between the child and the parent, which was the cause of this issue. I've added in a custom simple case of a circular reference as well 👍

File: tests/contrib/pytest/test_pytest.py [hunk 3]
```diff
@@ -122,19 +144,19 @@ def __init__(self, name, value):
             def item_param():
                 return 42
 
+            circular_reference = A("circular_reference", A("child", None))
+            circular_reference.value.value = circular_reference
```
Comment (Kyle-Verhoog) [inline tests/contrib/pytest/test_pytest.py:148]:
```suggestion
            circular_reference.value = circular_reference
```

i think

Comment (Yun-Kim) [reply tests/contrib/pytest/test_pytest.py:148]: Does this make more sense?
```python
parent = A("parent", None)
child = A("child", parent)
parent.value = child
```

Comment (Yun-Kim) [reply tests/contrib/pytest/test_pytest.py:148]: `circular_reference.value` is the child object, and the child object's value is the reference back to the parent, let me know if I should make that more clear!

Comment (Kyle-Verhoog) [reply tests/contrib/pytest/test_pytest.py:148]: ah yes! i didn't look closely enough at the `A` class

**Review by Kyle-Verhoog — DISMISSED** [review]

**Review by Kyle-Verhoog — DISMISSED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

**Review by Kyle-Verhoog — APPROVED** [review]

File: tests/contrib/pytest/test_pytest.py (hunk 2 above)
Comment (jd) [reply tests/contrib/pytest/test_pytest.py:157]: Oh ok! This was not obvious indeed.
Thank you!

**Review by jd — APPROVED** [review]

**Review by brettlangdon — APPROVED** [review]

**Comment by Kyle-Verhoog** [comment]
@mergifyio backport 0.50

**Pull request merged**